lib/client.ts

```diff
@@ -266,47 +311,32 @@ const requestWithFilters = (
     Promise.resolve(requestOptions)
   );
 
-  let response: ServiceClientResponse | null = null;
-  const responseThunk = () => response;
-
   return requestFilterPromise
-    .catch((err: Error) =>
-      wrapFailedError(client, ServiceClient.REQUEST_FILTER_FAILED, err)
-    )
+    .catch((error: Error) => {
+      throw new RequestFilterError(error, client.name);
+    })
     .then(paramsOrResponse =>
       paramsOrResponse instanceof ServiceClientResponse
         ? paramsOrResponse
-        : request(paramsOrResponse)
+        : request(paramsOrResponse).catch((error: RequestError) => {
```

Comment:
Hey! 

Without digging too deep into the code yet, could something like this be an option for error handling:
```js
request(paramsOrResponse).catch((error: RequestError) => {
    return a.b.c.d();
}).catch(ConnectionTimeoutError, RequestFailedError, function(e) {
    //If it is a ConnectionTimeoutError or RequestFailedError, will end up here because
    //it is a type error to reference property of undefined
}).catch(RequestUserTimeoutError, function(e) {
    //Will end up here if a was never declared at all
}).catch(function(e) {
    //Generic catch-the rest, error wasn't any of the above errors.
});
```

Yes, it might possibly convolute the implementation (or needs Bluebird as a dependency) or it might be hard to implement so that the backwards compatibility is still maintained. On the consumer side it would make the implementation super clean and AFAIK it should work with the `async/await` and `try/catch` block.

Reasoning, if/else checking would only remain on the happy path side of the call, and no such logic needed in the error handling.

Just a thought! Any opinions?


Example is strongly inspired by: http://bluebirdjs.com/docs/api/catch.html

Reply:
@pyryjook Thank you for your comment. I'm not sure though I completely understand what is the benefit of your code compared to just straight `instanceof` checks:

```js
request(paramsOrResponse).catch(e => {
  if (e instanceof ConnectionTimeoutError || e instanceof RequestFailedError) {
    //If it is a ConnectionTimeoutError or RequestFailedError, will end up here
  } else if (e instanceof RequestUserTimeoutError) {
    //Will end up here if a was never declared at all
  } else {
    //Generic catch-the rest, error wasn't any of the above errors.
  }
}
```

It is the same semantics and same amount of code but not using any non-standard promise features. Could you please clarify what makes your proposed solution cleaner? I feel like I'm missing something.

Especially not sure what you mean by this:

> It's a matter of semantics: If/else checking would only remain on the happy path side of the call, and no such logic needed in the error handling.

If you could provide more details  / examples that would really help.

Reply:
Yes, in practice both of the ways result the same end goal, that's for sure. The currently implemented change here is already a big improvement, so big thanks for that!

In many languages errors can be handled like this(Java):
```java
        try {
		File file = new File("./tmp.txt");
		inputStream = new FileInputStream(file);
		
		// use the inputStream to read a file
		
		// do NOT do this
		inputStream.close();
	} catch (FileNotFoundException e) {
		log.error(e);
	} catch (IOException e) {
		log.error(e);
	}
```

That clearly distinguishes the certain error handling to their respective `catch` block, instead of one `catch` with `if/else`. As the Bluebird documentation (link above) also states:

>  ...to work more like catch-clauses in languages like Java or C#. Instead of manually checking instanceof or .name === "SomeError", you may specify a number of error constructors which are eligible for this catch handler. The catch handler that is first met that has eligible constructors specified, is the one that will be called.

Semantics part for me here is that by doing it that way frees error handling from having such manual error checking, but instead errors would only be handled in respective `catch` block. In general it would be more of a pattern for error handling, then supported already by this library.

Hopefully this explains a bit more, what I meant here. I'm not saying that this would be easy to implement, just wanted to point out this possible and nice way of handling errors which is already implemented somewhere else.

Reply:
Yeah, maybe that is the best way to do it.